jzp-code/boot-demo/src/main/java/com/tutorial/boot_demo/service/StudentServiceImpl.java

```diff
@@ -0,0 +1,58 @@
+package com.tutorial.boot_demo.service;
+
+import com.fasterxml.jackson.databind.util.StdConverter;
+import com.tutorial.boot_demo.converter.StudentConverter;
+import com.tutorial.boot_demo.dao.Student;
+import com.tutorial.boot_demo.dao.StudentRepository;
+import com.tutorial.boot_demo.dto.StudentDTO;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import org.springframework.util.CollectionUtils;
+import org.springframework.util.StringUtils;
+
+import java.util.List;
+
+@Service
+public class StudentServiceImpl implements StudentService{
+
+    @Autowired
+    private StudentRepository studentRepository;
+
+    @Override
+    public StudentDTO getStudentById(long id) {
+        Student student = studentRepository.findById(id).orElseThrow(RuntimeException::new);
+        return StudentConverter.convertStudent(student);
+
+    }
+
+    @Override
+    public Long addNewStudent(StudentDTO studentDTO) {
+        List<Student> studentList = studentRepository.findByEmail(studentDTO.getEmail());
+        if (!CollectionUtils.isEmpty(studentList)) {
+            throw new IllegalStateException("email:" + studentDTO.getEmail() + " has been taken");
+        }

```

Comment:
> can we use table constraints instead of programming check? or at least we need adding an index on column `email`.

that helps, when design the table we can use "ADD CONSTRAINT UC_Email UNIQUE (email);" and “CREATE INDEX idx_email ON students(email);”.